astropy/nddata/utils.py

```diff
@@ -242,7 +245,15 @@ def extract_array(array_large, shape, position, mode='partial',
 
     # Extracting on the edges is presumably a rare case, so treat special here
     if (extracted_array.shape != shape) and (mode == 'partial'):
-        extracted_array = np.zeros(shape, dtype=array_large.dtype)
+        if np.isnan(fill_value):
+            dtype = float
+            warnings.warn('Output array dtype was changed to float to '
+                          'fill missing values with np.nan. Please '
+                          'change the fill_value keyword if this is not '
+                          'desirable.')
```

Comment:
And you need to catch it in the test too. 😉 
```suggestion
                          'desirable.', AstropyUserWarning)
```

Reply:
Ugh, there are now 10 failing tests, include some in the narrative docs.

Reply:
For the docs, you can use https://github.com/astropy/pytest-doctestplus#ignoring-warnings